modules/core/shared/src/main/scala/Schema.scala

```diff
@@ -246,13 +292,31 @@ object Schema {
     case object StrSet extends Schema[NonEmptySet[String]]
     case class Dictionary[A](value: Schema[A]) extends Schema[Map[String, A]]
     case class Sequence[A](value: Schema[A]) extends Schema[List[A]]
-    case class Record[R](value: FreeApplicative[Field[R, *], R])
-        extends Schema[R]
+
+    case class Record[R](
+        fields: List[Field[R, ?]],
+        build: List[Any] => R,
+        fieldNames: Array[String]
+    ) extends Schema[R]
+
+    object Record {
+      def apply[R](
+          fields: List[Field[R, ?]],
+          build: List[Any] => R
+      ): Record[R] = {
+        val fieldNames = fields.map(_.name).toArray
+        new Record(fields, build, fieldNames)
+      }
+    }
```

Comment:
I am not sure about this, I think we could replace with:
```
case class Record[R](or: OptimizedRecord[R,*]) extends Schema[R]
```

I don't believe caching fieldNames make this huge difference in the normal use cases